### PR TITLE
feat(gear-programs): Introduce `vft-manager`

### LINKED ISSUE #163: Merge implementations for `vft-gateway` and `vft-treasury`

### PR DESCRIPTION
**Pull request by mertwole**
Resolves #163

### CODE REVIEW

**Comment by ByteNacked** [comment]
Do we need e2e test like here: `gear-programs/vft-treasury/tests/end2end.rs` ?

File: gear-programs/vft-manager/app/src/services/mod.rs [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+        }
+    }
+
+    fn ensure_admin(&self) {
+        if self.state().admin != self.exec_context.actor_id() {
+            panic!("Not admin")
+        }
+    }
+
+    pub async fn submit_receipt(&mut self, receipt_rlp: Vec<u8>) -> Result<(), Error> {
```
Comment (ByteNacked) [inline gear-programs/vft-manager/app/src/services/mod.rs:186]: Perhaps add a comment for this method. After renaming, its functionality is not immediately obvious.

File: gear-programs/vft-manager/app/src/services/mod.rs [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+            TokenSupply::Ethereum => {
+                token_operations::mint(vara_token_id, receiver, amount, config, msg_id).await
+            }
+            TokenSupply::Gear => {
+                token_operations::unlock(vara_token_id, receiver, amount, config, msg_id).await
+            }
+        }
+    }
+
+    pub async fn request_bridging(
```
Comment (ByteNacked) [inline gear-programs/vft-manager/app/src/services/mod.rs:260]: Maybe comment here too?

File: gear-programs/vft-manager/app/src/services/mod.rs [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+        }
+    }
+}
+
+#[derive(Debug, Decode, Encode, TypeInfo, Clone)]
+pub struct Config {
+    gas_to_burn_tokens: u64,
+    gas_for_reply_deposit: u64,
+    gas_to_mint_tokens: u64,
+    gas_to_transfer_tokens: u64,
```
Comment (ByteNacked) [inline gear-programs/vft-manager/app/src/services/mod.rs:89]: Do we really need both? Maybe just `gas_to_token_tx` or something like that.

File: gear-programs/vft-manager/app/src/services/mod.rs [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+    gas_to_process_mint_request: u64,
```
Comment (ByteNacked) [inline gear-programs/vft-manager/app/src/services/mod.rs:90]: Shouldn't it be `gas_to_process_mint_request` -> `gas_to_submit_receipt` ?

File: gear-programs/vft-manager/app/src/services/mod.rs [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+    gas_to_send_request_to_builtin: u64,
+    reply_timeout: u32,
+    gas_for_transfer_to_eth_msg: u64,
```
Comment (ByteNacked) [inline gear-programs/vft-manager/app/src/services/mod.rs:93]: `gas_for_transfer_to_eth_msg` -> `gas_for_request_bridging` ?

**Comment by ByteNacked** [comment]
GJ :+1: :+1: :+1:

File: gear-programs/vft-manager/app/src/services/error.rs [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+pub enum Error {
+    SendFailure,
+    ReplyFailure,
+    BurnTokensDecode,
+    TransferFromDecode,
+    BurnFailure,
+    RequestToBuiltinSend,
+    RequestToBuiltinReply,
+    BuiltinDecode,
+    PayloadSize,
```
Comment (ByteNacked) [inline gear-programs/vft-manager/app/src/services/error.rs:13]: Seems not used

File: gear-programs/vft-manager/app/src/services/error.rs [hunk 7]
```diff
@@ -0,0 +1,30 @@
+use sails_rs::prelude::*;
+
+#[derive(Debug, Encode, Decode, TypeInfo, Clone, PartialEq, Eq)]
+pub enum Error {
+    SendFailure,
+    ReplyFailure,
+    BurnTokensDecode,
+    TransferFromDecode,
+    BurnFailure,
```
Comment (ByteNacked) [inline gear-programs/vft-manager/app/src/services/error.rs:9]: Seems not used

**Comment by mertwole** [comment]
> Do we need e2e test like here: `gear-programs/vft-treasury/tests/end2end.rs` ?

What else you propose to add? I think current tests cover every use-case possible: bridging to/from ethereum * supply on gear/ethereum

File: gear-programs/vft-manager/app/src/services/ERC20Manager.json [hunk 8]
```diff
@@ -0,0 +1 @@
...0921","dweb:/ipfs/QmQQCDMcqEasWJymxzsNL7gS2iKZAic67VqcHPaUHfQRwC"],"license":null},"src/ERC20GearSupply.sol":{"keccak256":"0xb58456aa3f0ee9a2bd9eb262f0bc7865f03ecf70089dccf955c74fe2265bde83","urls":["bzz-raw://87546639b968be049ed4b56a2955e238c609c01276341f364b47cb39fd5bf5c0","dweb:/ipfs/QmNPWEYVQi9SCZ6xUPMhQAcRxirXno1JozFy9KDshMVA4v"],"license":null},"src/ERC20Manager.sol":{"keccak256":"0x338a5ffbca7c77a9129c65dd5c5a385bcbc6d09669670681cdc88cc6e0c8a160","urls":["bzz-raw://893ef20017e7d1a70db30dcb75cc5900093caf722c33c986f6207af2bd714519","dweb:/ipfs/QmdVUPXqmC446WfDM2Chj8rdkSt1nJnJd5v6E27Z8QWvqQ"],"license":null},"src/interfaces/IERC20Manager.sol":{"keccak256":"0x7da0697558168d69b5fbc2380203c11dfee76c8734ab0f512b90879be5b02711","urls":["bzz-raw://ce00c2bea23f0a16ff8e27e8b22d0a1d776799e9f316009dc47482f26827d1c1","dweb:/ipfs/QmYz2SYxPA6tNBDYchPnBnWsB9seUVHqXykDieusKzpVz8"],"license":null},"src/interfaces/IMessageQueue.sol":{"keccak256":"0xf8352567197fd1fe695309bfbb2416c79d76e301cd267215476cd94a14784e74","urls":["bzz-raw://ed0f4f31af4e2a09b9bf190f07b2ead5495ea8b61a0e35bc48d161282e8bc015","dweb:/ipfs/QmP164SwLhsAsAbBPkQm5iQqmgfAcSvfhFpEov4eGRhfD9"],"license":null}},"version":1},"id":54}
```
[the remark this answers is not in the document]
Comment (mertwole) [reply gear-programs/vft-manager/app/src/services/ERC20Manager.json:1]: Introduced `build.rs` script. It should work when https://github.com/gear-tech/gear-bridges/pull/194 will be merged

**Review by ByteNacked — APPROVED** [review]

File: gear-programs/vft-manager/app/build.rs [hunk 9]
```diff
@@ -0,0 +1,20 @@
+use std::{env::current_dir, process::Command};
+
+fn main() {
+    let mut ethereum_contracts_dir = current_dir().expect("Failed to get current dir");
```
Comment (gshep) [inline gear-programs/vft-manager/app/build.rs:4]: IMO it's better to depend on the specific environment variable:
```
 let idl_file_path = {
        let mut path = PathBuf::from(env::var("CARGO_MANIFEST_DIR").unwrap());
        path.pop();
        path.pop();

        path.push("vft-gateway");
        path.push("src");
        path.push("wasm");
        path.push("vft-gateway.idl");

        path
    };
```

File: gear-programs/vft-manager/app/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,19 @@
+#![no_std]
+
+use sails_rs::{gstd::GStdExecContext, prelude::*};
+pub mod services;
+use services::{InitConfig, VftManager};
+#[derive(Default)]
```
Comment (gshep) [inline gear-programs/vft-manager/app/src/lib.rs:6]: we definitely need the empty line here )

File: gear-programs/vft-manager/app/src/services/bridge_builtin_operations.rs [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+#[derive(Debug, Decode, Encode, TypeInfo)]
+pub struct Payload {
+    pub receiver: H160,
+    pub token_id: H160,
+    pub amount: U256,
+}
+
+impl Payload {
+    pub fn pack(self) -> Vec<u8> {
+        let mut packed = Vec::with_capacity(20 + 20 + 32); // H160 is 20 bytes, U256 is 32 bytes
```
Comment (gshep) [inline gear-programs/vft-manager/app/src/services/bridge_builtin_operations.rs:44]: if i'm not mistaken there are no comments aside the code. Usually it places above(before) the next expression.

**Review by gshep — APPROVED** [review]
:+1:

**Pull request merged**